common/src/jni/main/cpp/NativeCrypto.cpp

```diff
@@ -7316,16 +7316,93 @@ static jint NativeCrypto_SSL_read(JNIEnv* env, jclass, jlong ssl_address, jobjec
+        JNI_TRACE("ssl=%p NativeCrypto_SSL_read => b == null", ssl);
+        return 0;
+    }
 
-    ScopedByteArrayRW bytes(env, b);
-    if (bytes.get() == nullptr) {
-        JNI_TRACE("ssl=%p NativeCrypto_SSL_read => threw exception", ssl);
+    size_t array_size = static_cast<size_t>(env->GetArrayLength(b));
+    if (ARRAY_CHUNK_INVALID(array_size, offset, len)) {
+        Errors::jniThrowException(env, "java/lang/ArrayIndexOutOfBoundsException", "b");
+        JNI_TRACE("ssl=%p NativeCrypto_SSL_read => ArrayIndexOutOfBoundsException", ssl);
         return 0;
     }
 
     OpenSslError sslError;
-    int ret = sslRead(env, ssl, fdObject, shc, reinterpret_cast<char*>(bytes.get() + offset), len,
+    int ret;
+    if (JniUtil::isGetByteArrayElementsLikelyToReturnACopy(array_size)) {
+        if (len <= 1024) {
+            // Allocate small buffers on the stack for performance.
+            jbyte buf[1024];
+            ret = sslRead(env, ssl, fdObject, shc, reinterpret_cast<char*>(&buf[0]), len, sslError,
+                          read_timeout_millis);
+            if (ret > 0) {
+                // Don't bother applying changes if issues were encountered.
+                env->SetByteArrayRegion(b, offset, ret, &buf[0]);
+            }
+        } else {
+            // Allocate larger buffers on the heap.
+            // TODO: Enforce that len >= 0? If not, the cast below could result in a memory
+            // allocation error/attack.
+            jint remaining = len;
+            jint buf_size = (remaining >= 65536) ? 65536 : remaining;
+            std::unique_ptr<jbyte[]> buf(new jbyte[static_cast<unsigned int>(buf_size)]);
+            // TODO: Doesn't the following condition require new(std::nothrow)?
+            if (buf.get() == nullptr) {
+                Errors::jniThrowOutOfMemory(env, "Unable to allocate chunk buffer");
+                return 0;
+            }
+            jint expected_reads = len / buf_size;
+            if (expected_reads > 0) {
+                // TODO: Fix cumulative read timeout? This currently ensures that the cumulative
+                // timeout does not exceed the given timeout, but it is overly strict in that it
+                // allocates uniformly-sized time slots for each call.
+                read_timeout_millis = read_timeout_millis / expected_reads;
+            }
+            ret = 0;
+            while (remaining > 0) {
+                jint temp_ret;
+                jint chunk_size = (remaining >= buf_size) ? buf_size : remaining;
+                // TODO: Make sure it's safe to reuse same callbacks repeatedly.
+                temp_ret = sslRead(env, ssl, fdObject, shc, reinterpret_cast<char*>(buf.get()),
+                                   chunk_size, sslError, read_timeout_millis);
+                if ((temp_ret == THROWN_EXCEPTION || temp_ret == -1) && ret > 0) {
+                    // TODO: Should we only check for -1 above?
```

Comment:
Is there a possibility that THROWN_EXCEPTION ever refers to a different exception that _won't_ be thrown on the next call? If so, then it might not be safe to throw it away.